### PR TITLE
[eas-cli] Improve EAS metadata categories configuration

### PR DESCRIPTION
**Pull request by byCedric**
# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

This improves/finalizes the `apple.categories` in the EAS metadata store configuration.

> See the `metadata-0.json` schema for examples

# How

- Updated schema documentation and validation
- Updated the schema typescript types
- Updated the store configuration reader and writer

# Test Plan

- See added tests

### CODE REVIEW

File: packages/eas-cli/src/metadata/apple/config/writer.ts [hunk 1]
```diff
@@ -64,15 +66,53 @@ export class AppleConfigWriter {
...
-      if (secondaryCategory) {
-        this.schema.categories.push(secondaryCategory.id as any);
+    // We are receiving strings, but need AppCategoryId or AppSubcategoryId.
+    // These are returned by the ASC API and should be valid categories.
+    const primaryCat = attributes.primaryCategory.id as AppCategoryId;
+    const primarySubOne = attributes.primarySubcategoryOne?.id as AppSubcategoryId | undefined;
+    const primarySubTwo = attributes.primarySubcategoryTwo?.id as AppSubcategoryId | undefined;
+    const secondaryCat = attributes.secondaryCategory?.id as AppCategoryId | undefined;
+    const secondarySubOne = attributes.secondarySubcategoryOne?.id as AppSubcategoryId | undefined;
+    const secondarySubTwo = attributes.secondarySubcategoryTwo?.id as AppSubcategoryId | undefined;
```
Comment (byCedric) [inline packages/eas-cli/src/metadata/apple/config/writer.ts:92]: I kinda hate doing this, but the types are strict. Our model has `string` types for each attribute, but our schema types have `AppCategoryId | AppSubcategoryId`.

Alternatively, we could also relax the schema type to just match `string`. This is validated anyways in the schema :) cc @EvanBacon

Comment (byCedric) [reply packages/eas-cli/src/metadata/apple/config/writer.ts:92]: nvm, I'm going to relax the types.

**Review by EvanBacon — APPROVED** [review]

File: packages/eas-cli/schema/metadata-0.json [hunk 2]
```diff
@@ -33,69 +33,154 @@
...
+            },
+            {
+              "type": "array",
+              "description": "The GAMES category can have additional subcategories",
+              "minItems": 1,
+              "maxItems": 3,
+              "uniqueItems": true,
+              "items": [
+                {
+                  "const": "GAMES"
```
Comment (EvanBacon) [inline packages/eas-cli/schema/metadata-0.json:79]: clever

**Comment by byCedric** [comment]
/changelog-entry chore Improve store config categories configuration

File: packages/eas-cli/src/metadata/apple/config/__tests__/fixtures/appInfo.ts [hunk 3]
```diff
@@ -1,23 +1,37 @@
...
+ * adding them here as fixed objects won't be readable.
+ */
+export function makeCategory(id?: AppCategoryId | AppSubcategoryId): AppCategory | undefined {
+  if (!id) {
+    return undefined;
+  }
 
-export const secondaryOnlyCategory: Pick<AttributesOf<AppInfo>, 'secondaryCategory'> = {
-  secondaryCategory: new AppCategory({} as any, 'GAMES', {
+  return new AppCategory({} as any, id, {
```
Comment (dsokal) [inline packages/eas-cli/src/metadata/apple/config/__tests__/fixtures/appInfo.ts:18]: Don't use `as any`

File: packages/eas-cli/src/metadata/apple/config/reader.ts [hunk 4]
```diff
@@ -78,14 +78,33 @@ export class AppleConfigReader {
...
+      categoryIds.secondaryCategory = categories[1][0];
+      categoryIds.secondarySubcategoryOne = categories[1][1];
+      categoryIds.secondarySubcategoryTwo = categories[1][2];
+    } else {
+      categoryIds.secondaryCategory = categories[1];
+    }
+
+    // Because we handle categories as normal strings,
+    // the type doesn't match with the actual CategoryIds types.
+    return categoryIds as CategoryIds;
```
Comment (dsokal) [inline packages/eas-cli/src/metadata/apple/config/reader.ts:107]: I don't get it, if categories are strings why would we want to pretend they are objects? Maybe we should alter the type?

File: packages/eas-cli/src/metadata/apple/config/writer.ts [hunk 5]
```diff
@@ -64,15 +64,44 @@ export class AppleConfigWriter {
...
+      // autofill with an empty string and cause a store config error.
+      this.schema.categories[0] = attributes.primaryCategory?.id ?? '';
+    }
+
+    if (attributes.secondaryCategory && attributes.secondarySubcategoryOne) {
+      this.schema.categories[1] = [
+        attributes.secondaryCategory.id,
+        attributes.secondarySubcategoryOne?.id,
+        attributes.secondarySubcategoryTwo?.id,
+      ].filter(Boolean) as string[];
```
Comment (dsokal) [inline packages/eas-cli/src/metadata/apple/config/writer.ts:102]: 1. I guess you're trying to filter falsy values here. Unfortunately, filtering like this is not correct. See:
  Screenshot 2022-07-04 at 09 15 54
2. It should be sth like:

```
const arr = [
    undefined,
    'string',
    null,
]

const isString = (value: string | undefined | null): value is string => typeof value == 'string';
const filtered = arr.filter(isString)
```

**Pull request merged**